File: alien/Go/src-std/CountXmlNodes.go
package main

import (
	"encoding/xml"
	"fmt"
	"io"
	"os"
)

func countNodeOccurrences(xmlFilePath string, filterNodes map[string]struct{}) (map[string]int, error) {
	file, err := os.Open(xmlFilePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	decoder := xml.NewDecoder(file)
	nodeCounts := make(map[string]int)

	for {
		tok, err := decoder.Token()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}

		switch se := tok.(type) {
		case xml.StartElement:
			nodeName := se.Name.Local
			if len(filterNodes) == 0 {
				nodeCounts[nodeName]++
			} else {
				if _, ok := filterNodes[nodeName]; ok {
					nodeCounts[nodeName]++
				}
			}
		}
	}
	return nodeCounts, nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <path_to_xml_file> [node_name1 node_name2 ...]\n", os.Args[0])
		os.Exit(1)
	}

	xmlFilePath := os.Args[1]
	filterNodes := make(map[string]struct{})
	if len(os.Args) > 2 {
		for _, name := range os.Args[2:] {
			filterNodes[name] = struct{}{}
		}
	}

	nodeCounts, err := countNodeOccurrences(xmlFilePath, filterNodes)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error processing XML file: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Node counts:")
	for node, count := range nodeCounts {
		fmt.Printf("%s: %d\n", node, count)
	}
}

File: alien/Julia/LightXML/CountXmlNodes.jl
using LightXML


function traverse(node::XMLElement, node_counts::Dict{String, Int}, filter_nodes::Union{Nothing, Set{String}}=nothing)
        node_name = LightXML.name(node)
        if isnothing(filter_nodes) || node_name in filter_nodes
            node_counts[node_name] = get(node_counts, node_name, 0) + 1
        end
        for child in LightXML.child_elements(node)
            traverse(child, node_counts, filter_nodes)
        end
    end


function count_node_occurrences(xml_file_path::String, filter_nodes::Union{Nothing, Set{String}}=nothing)
    doc = parse_file(xml_file_path)
    node_counts = Dict{String, Int}()

    traverse(root(doc), node_counts, filter_nodes)
    return node_counts
end


function main()
    if length(ARGS) < 1
        println("Usage: julia CountXmlNodes.jl <path_to_xml_file> [node_name1 node_name2 ...]")
        exit(1)
    end

    xml_file_path = ARGS[1]
    filter_nodes = length(ARGS) > 1 ? Set(ARGS[2:end]) : nothing

    try
        node_counts = count_node_occurrences(xml_file_path, filter_nodes)
        println("Node counts:")
        for (node, count) in node_counts
            println("$(node): $(count)")
        end
    catch e
        println("Error processing XML file: $e")
        exit(1)
    end
end

main()